streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java

```diff
@@ -189,8 +195,8 @@ public void processInOrder(final K key, final V value, final long timestamp) {
             //create left window for new record
             if (!leftWinAlreadyCreated) {
                 final ValueAndTimestamp<Agg> valueAndTime;
-                //there's a right window that the new record could create --> new record's left window is not empty
-                if (latestLeftTypeWindow != null) {
+                // if there's a right window that the new record could create --> new record's left window is not empty
```

Comment:
Now that I think about it, this isn't exactly true -- you could have the previous record at 10 and the current record at 21 (size is 10), then you'd have to create a right window for [11, 21] but the left window is also [11, 21] which is empty (except for the current record of course).So you'd want to check that `previousRecord >= currentRecord - timeDifference` which is actually technically what `previousRightWindowPossible(rightWindow, timestamp)` returns but it took me a long time to figure all that out because we're sort of passing in the wrong parameters here. In this call we pass in `latestLeftTypeWindow.value.timestamp` which is equivalent to `previousRecord.timestamp` which is not conceptually the same as the `rightWindowStart` . I know the logic works out the way you intended but it's pretty hard to untangle. You should update the comment and create a separate method for this case which you can name more accurately for this specific scenario

Reply:
It actually still works for that example. The range doesn't pick up the window from [0,10] as it goes from timestamp - 2*timeDifference, so for this example the range would start at 1. I initially did this to avoid these scenarios entirely. You are right though that the comment doesn't hold for the combined window scenarios. For example, we find the combined window and store it as "latestLeftTypeWindow" when we process a record at say, 14. If the max record in "latestLeftTypeWindow" is 2, then we don't want to create the right window for that, since it would be [3,13].

I improved the comment to be `if there's a right window that the new record could create && max record falls within left window --> new record's left window is not empty`

Reply:
Here too, is this kind of moot now that we can just track the `previousRecordTimestamp`? IIUC all we really want to do is make sure that the left window is not empty, which is actually a pretty simple calculation in terms of the `previousRecordTimestamp`